registration/include/pcl/registration/ndt_omp.h

```diff
@@ -0,0 +1,521 @@
+   */
+  virtual void
+  computeTransformation(PointCloudSource& output)
+  {
+    computeTransformation(output, Eigen::Matrix4f::Identity());
+  }
+
+  /** \brief Estimate the transformation and returns the transformed source (input) as output.
+   * \param[out] output the resultant input transformed point cloud dataset
+   * \param[in] guess the initial gross estimation of the transformation
+   */
+  void
+  computeTransformation(PointCloudSource& output, const Eigen::Matrix4f& guess) override;
+
+  /** \brief Initiate covariance voxel structure. */
+  void inline
+  init()
+  {
+    target_cells_.setLeafSize(resolution_, resolution_, resolution_);
+    target_cells_.setInputCloud(target_);
+    // Initiate voxel structure.
+    target_cells_.filter(true);
+  }
+
+  /** \brief Compute derivatives of probability function w.r.t. the transformation vector.
+    * \note Equation 6.10, 6.12 and 6.13 [Magnusson 2009].
+    * \param[out] score_gradient the gradient vector of the probability function w.r.t. the transformation vector
+    * \param[out] hessian the hessian matrix of the probability function w.r.t. the transformation vector
+    * \param[in] trans_cloud transformed point cloud
+    * \param[in] p the current transform vector
+    * \param[in] compute_hessian flag to calculate hessian, unnessissary for step calculation.
+    */
+  double
+  computeDerivatives(Eigen::Matrix<double, 6, 1>& score_gradient,
+                     Eigen::Matrix<double, 6, 6>& hessian,
+                     PointCloudSource& trans_cloud,
+                     Eigen::Matrix<double, 6, 1>& p,
+                     bool compute_hessian = true);
+
+  /** \brief Compute individual point contirbutions to derivatives of probability function w.r.t. the transformation vector.
+    * \note Equation 6.10, 6.12 and 6.13 [Magnusson 2009].
+    * \param[in,out] score_gradient the gradient vector of the probability function w.r.t. the transformation vector
+    * \param[in,out] hessian the hessian matrix of the probability function w.r.t. the transformation vector
+    * \param[in] x_trans transformed point minus mean of occupied covariance voxel
+    * \param[in] c_inv covariance of occupied covariance voxel
+    * \param[in] compute_hessian flag to calculate hessian, unnessissary for step calculation.
+    */
+  double
+  updateDerivatives(Eigen::Matrix<double, 6, 1>& score_gradient,
+                    Eigen::Matrix<double, 6, 6>& hessian,
+                    const Eigen::Matrix<float, 4, 6>& point_gradient,
+                    const Eigen::Matrix<float, 24, 6>& point_hessian,
+                    const Eigen::Vector3d& x_trans,
+                    const Eigen::Matrix3d& c_inv,
+                    bool compute_hessian = true) const;
+
+  /** \brief Precompute anglular components of derivatives.
+   * \note Equation 6.19 and 6.21 [Magnusson 2009].
+   * \param[in] p the current transform vector
```

Comment:
`p` isn't a nice name for a transform vector in implementation, although it may be true to its literary roots